File: jvm/core/main/scala/com/webank/eggroll/core/resourcemanager/job/container/WorkingDirectoryPreparer.scala
package com.webank.eggroll.core.resourcemanager.job.container

import java.io.{ByteArrayInputStream, FileOutputStream, InputStream}
import java.nio.file.Path
import java.util.zip.{ZipEntry, ZipInputStream}

class WorkingDirectoryPreparer(
                                var files: Map[String, Array[Byte]],
                                var zippedFiles: Map[String, Array[Byte]],
                                var need_cleanup: Boolean = true,
                                var workingDir: Path = null
                              ) {

  def setWorkingDir(workingDir: Path): Unit = {
    this.workingDir = workingDir
  }

  def setNeedCleanup(need_cleanup: Boolean): Unit = {
    this.need_cleanup = need_cleanup
  }

  private def getWorkingDir(): Path = {
    if (workingDir != null) {
      workingDir
    } else {
      throw new Exception("workingDir is not set")
    }
  }

  def prepare(): Unit = {
    getWorkingDir().toFile.mkdirs()
    files.foreach { case (fileName, content) =>
      val file = getWorkingDir().resolve(fileName).toFile
      val fos = new FileOutputStream(file)
      fos.write(content)
      fos.close()
    }
    zippedFiles.foreach { case (dirname, zipBytes) =>
      val zipInputStream = new ZipInputStream(new ByteArrayInputStream(zipBytes))
      val buffer = new Array[Byte](1024)

      def extractEntry(entry: ZipEntry, inputStream: InputStream, outputPath: Path): Unit = {
        val file = outputPath.toFile
        file.getParentFile.mkdirs()
        val fileOutputStream = new FileOutputStream(file)

        try {
          var bytesRead = inputStream.read(buffer)
          while (bytesRead != -1) {
            fileOutputStream.write(buffer, 0, bytesRead)
            bytesRead = inputStream.read(buffer)
          }
        } finally {
          fileOutputStream.close()
        }
      }

      Iterator.continually(zipInputStream.getNextEntry)
        .takeWhile(_ != null)
        .foreach(entry => {
          val outputPath = getWorkingDir().resolve(dirname).resolve(entry.getName)
          if (!entry.isDirectory) {
            extractEntry(entry, zipInputStream, outputPath)
          }
          zipInputStream.closeEntry()
        })

      zipInputStream.close()
    }
    // clear files and zippedFiles
    files = Map.empty
    zippedFiles = Map.empty
  }

  def cleanup(): Unit = {
    if (need_cleanup && workingDir != null) {
      workingDir.toFile.delete()
    }
  }
}

File: jvm/core/src/main/java/org/fedai/eggroll/core/containers/container/WorkingDirectoryPreparer.java
package org.fedai.eggroll.core.containers.container;


import lombok.Data;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class WorkingDirectoryPreparer {
    private Map<String, byte[]> files = new HashMap<>();
    private Map<String, byte[]> zippedFiles = new HashMap<>();
    ;
    private boolean need_cleanup = true;
    private Path workingDir = null;

    private Path getWorkingDir() throws Exception {
        if (workingDir != null) {
            return workingDir.toAbsolutePath();
        } else {
            throw new Exception("workingDir is not set");
        }
    }


    private Path getModelsDir() throws Exception {
        return getWorkingDir().resolve("models");
    }

    private Path getLogsDir() throws Exception {
        return getWorkingDir().resolve("logs");
    }

    public Map<String, String> getContainerDirEnv() throws Exception {
        String workingDirString = getWorkingDir().toString();
        String modelsDirString = getModelsDir().toString();
        String logsDirString = getLogsDir().toString();

        Map<String, String> envMap = new HashMap<>();
        envMap.put("EGGROLL_CONTAINER_DIR", workingDirString);
        envMap.put("EGGROLL_CONTAINER_MODELS_DIR", modelsDirString);
        envMap.put("EGGROLL_CONTAINER_LOGS_DIR", logsDirString);

        return envMap;
    }

    public void prepare() {
        try {
            Files.createDirectories(getWorkingDir());
            Files.createDirectories(getModelsDir());
            Files.createDirectories(getLogsDir());

            for (Map.Entry<String, byte[]> entry : files.entrySet()) {
                String fileName = entry.getKey();
                byte[] content = entry.getValue();

                Path filePath = getWorkingDir().resolve(fileName);
                File file = filePath.toFile();

                FileOutputStream fos = new FileOutputStream(file);
                fos.write(content);
                fos.close();
            }

            for (Map.Entry<String, byte[]> entry : zippedFiles.entrySet()) {
                String dirname = entry.getKey();
                byte[] zipBytes = entry.getValue();
                ZipInputStream zipInputStream = new ZipInputStream(new ByteArrayInputStream(zipBytes));
                byte[] buffer = new byte[1024];

                ZipEntry zipEntry;
                while ((zipEntry = zipInputStream.getNextEntry()) != null) {
                    Path outputPath = getWorkingDir().resolve(dirname).resolve(zipEntry.getName());

                    if (!zipEntry.isDirectory()) {
                        File outputFile = outputPath.toFile();
                        outputFile.getParentFile().mkdirs();

                        FileOutputStream fileOutputStream = new FileOutputStream(outputFile);
                        int bytesRead;
                        while ((bytesRead = zipInputStream.read(buffer)) != -1) {
                            fileOutputStream.write(buffer, 0, bytesRead);
                        }
                        fileOutputStream.close();
                    }

                    zipInputStream.closeEntry();
                }

                zipInputStream.close();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        files.clear();
        zippedFiles.clear();
    }


    public void cleanup() throws Exception {
        if (need_cleanup && workingDir != null) {
            File workingDirFile = workingDir.toFile();
            workingDirFile.delete();
        }
    }

    public Map<String, byte[]> getFiles() {
        return files;
    }

    public void setFiles(Map<String, byte[]> files) {
        this.files = files;
    }

    public Map<String, byte[]> getZippedFiles() {
        return zippedFiles;
    }

    public void setZippedFiles(Map<String, byte[]> zippedFiles) {
        this.zippedFiles = zippedFiles;
    }

    public boolean isNeed_cleanup() {
        return need_cleanup;
    }

    public void setNeed_cleanup(boolean need_cleanup) {
        this.need_cleanup = need_cleanup;
    }

    public void setWorkingDir(Path workingDir) {
        this.workingDir = workingDir;
    }
}
